File: ios/YamabikoChat/Data/Models/AutoConversation.swift
import Foundation
import GRDB

enum AutoConversationStatus: String, Codable, CaseIterable {
    case active = "ACTIVE"
    case paused = "PAUSED"
    case ended = "ENDED"
}

enum AutoConversationEndReason {
    static let userStop = "USER_STOP"
    static let maxTurns = "MAX_TURNS"
    static let endSignal = "END_SIGNAL"
    static let error = "ERROR"
    static let apiError = "API_ERROR"
}

struct AutoConversationConfig {
    var title: String
    var modelA: String
    var modelB: String
    var providerA: String
    var providerB: String
    var systemPromptA: String
    var systemPromptB: String
    var maxTurns: Int
    var endSignal: String
}

struct AutoConversation: Codable, FetchableRecord, MutablePersistableRecord, Identifiable, Equatable {
    static let databaseTableName = "auto_conversations"

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case modelA
        case modelB
        case providerA
        case providerB
        case systemPromptA
        case systemPromptB
        case statusRaw = "status"
        case maxTurns
        case currentTurn
        case createdAtMs
        case lastActiveAtMs
        case endReason
        case endSignal
        case boundChatConversationId
    }

    var id: Int64?
    var title: String
    var modelA: String
    var modelB: String
    var providerA: String
    var providerB: String
    var systemPromptA: String
    var systemPromptB: String
    var statusRaw: String
    var maxTurns: Int
    var currentTurn: Int
    var createdAtMs: Int64
    var lastActiveAtMs: Int64
    var endReason: String?
    var endSignal: String
    var boundChatConversationId: Int64?

    var status: AutoConversationStatus {
        get { AutoConversationStatus(rawValue: statusRaw.uppercased()) ?? .ended }
        set { statusRaw = newValue.rawValue }
    }

    init(
        id: Int64? = nil,
        title: String,
        modelA: String,
        modelB: String,
        providerA: String,
        providerB: String,
        systemPromptA: String,
        systemPromptB: String,
        status: AutoConversationStatus = .active,
        maxTurns: Int,
        currentTurn: Int = 0,
        createdAtMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        lastActiveAtMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        endReason: String? = nil,
        endSignal: String = "[END]",
        boundChatConversationId: Int64? = nil
    ) {
        self.id = id
        self.title = title
        self.modelA = modelA
        self.modelB = modelB
        self.providerA = providerA
        self.providerB = providerB
        self.systemPromptA = systemPromptA
        self.systemPromptB = systemPromptB
        statusRaw = status.rawValue
        self.maxTurns = maxTurns
        self.currentTurn = currentTurn
        self.createdAtMs = createdAtMs
        self.lastActiveAtMs = lastActiveAtMs
        self.endReason = endReason
        self.endSignal = endSignal
        self.boundChatConversationId = boundChatConversationId
    }

    mutating func didInsert(_ inserted: InsertionSuccess) {
        id = inserted.rowID
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/local/AutoConversation.kt
package com.porarri.yamabikochat.data.local

import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity(tableName = "auto_conversations")
data class AutoConversation(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val title: String,
    val modelA: String,
    val modelB: String,
    val providerA: String,
    val providerB: String,
    val systemPromptA: String,
    val systemPromptB: String,
    val status: AutoConversationStatus,
    val maxTurns: Int = 20, // デフォルトは20ターン
    val currentTurn: Int = 0,
    val createdAt: Long = System.currentTimeMillis(),
    val lastActiveAt: Long = System.currentTimeMillis(),
    val endReason: String? = null,
    val endSignal: String = "[END]",
    val boundChatConversationId: Long? = null
)

enum class AutoConversationStatus {
    ACTIVE,    // 進行中
    PAUSED,    // 一時停止
    ENDED      // 終了
}

@Entity(tableName = "auto_conversation_messages")
data class AutoConversationMessage(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val autoConversationId: Long,
    val turnNumber: Int,
    val speakerModel: String, // "A" or "B"
    val content: String,
    val reasoning: String? = null,
    val timestamp: Long = System.currentTimeMillis(),
    val isEndSignal: Boolean = false // 終了シグナルが含まれているか
)

// 会話の設定情報をまとめたデータクラス
data class AutoConversationConfig(
    val title: String,
    val modelA: String,
    val modelB: String,
    val providerA: String,
    val providerB: String,
    val systemPromptA: String,
    val systemPromptB: String,
    val maxTurns: Int = 20,
    val endSignal: String = "[END]" // 終了シグナルのキーワード
)

// 会話の詳細情報（メッセージ付き）
data class FullAutoConversation(
    val conversation: AutoConversation,
    val messages: List<AutoConversationMessage>
)

// 終了理由の定数
object AutoConversationEndReason {
    const val USER_STOP = "USER_STOP"           // ユーザーが手動停止
    const val MAX_TURNS = "MAX_TURNS"           // 最大ターン数到達
    const val END_SIGNAL = "END_SIGNAL"         // 終了シグナル検出
    const val ERROR = "ERROR"                   // エラーによる停止
    const val TIMEOUT = "TIMEOUT"               // タイムアウト
    const val API_ERROR = "API_ERROR"           // API エラー
}